### PR TITLE
Fix/sas initiator

### PR DESCRIPTION
**Pull request by David-T-White**

### CODE REVIEW

File: pkg/controller/provisioner.go [hunk 1]
```diff
@@ -23,10 +24,27 @@ var (
 	}
 )
 
+// Extract available SAS addresses for the Node from topology segments
+func parseTopology(topology []*csi.Topology, parameters *map[string]string) error {
+	klog.Infof("Topology: %v", topology)
+
+	for _, topo := range topology {
+		for key, val := range topo.GetSegments() {
+			if strings.Contains(key, "seagate-exos-x-csi/sas-address") {
```
Comment (jskazinski) [inline pkg/controller/provisioner.go:33]:
```suggestion
			if strings.Contains(key, fmt.Sprintf(%s/%s", common.TopologyPrefix, common.SasLabel) {
```

**Review by jskazinski — CHANGES_REQUESTED** [review]

File: pkg/node/node.go [hunk 2]
```diff
@@ -103,9 +106,24 @@ func (node *Node) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest)
...
 
+	topology := map[string]string{}
+	sasAddresses, err := getSASAddresses()
+	if err != nil {
+		klog.Infof("Error while searching for SAS HBA Addresses: %s", err)
+	}
+
+	for i, sasAddr := range sasAddresses {
+		//maximum value length 63 chars
+		topoKey := common.TopologyPrefix + "/" + "sas-address-" + strconv.Itoa(i)
```
Comment (jskazinski) [inline pkg/node/node.go:117]: Perhaps a string length test and warning if the string is longer than 63 characters?

**Review by jskazinski — CHANGES_REQUESTED** [review]

File: pkg/controller/publisher.go [hunk 3]
```diff
@@ -21,12 +71,28 @@ func (driver *Controller) ControllerPublishVolume(ctx context.Context, req *csi.
...
+	parameters := req.GetVolumeContext()
 
 	volumeName, _ := common.VolumeIdGetName(req.GetVolumeId())
-	initiatorName := req.GetNodeId()
-	klog.Infof("attach request for initiator %s, volume id: %s", initiatorName, volumeName)
 
-	lun, err := driver.client.PublishVolume(volumeName, initiatorName)
+	var initiatorNames []string
+	// Available SAS initiators for the node are provided here through NodeGetInfo
+	if parameters[common.StorageProtocolKey] == common.StorageProtocolSAS {
```
[the remark this answers is not in the document]
Comment (David-T-White) [reply pkg/controller/publisher.go:80]: I'm planning to do this separately soon

File: pkg/controller/publisher.go [hunk 4]
```diff
@@ -44,17 +110,36 @@ func (driver *Controller) ControllerUnpublishVolume(ctx context.Context, req *cs
...
 
-		return nil, err
+	klog.Infof("unmapping volume %s from initiator %s", volumeName, initiators)
+	for _, initiator := range initiators {
+		//TODO: Optimize for minimal API calls
+		_, status, err := driver.client.UnmapVolume(volumeName, initiator)
+		if err != nil {
+			if status != nil && status.ReturnCode == unmapFailedErrorCode {
+				klog.Info("unmap failed, assuming volume is already unmapped")
+				return &csi.ControllerUnpublishVolumeResponse{}, nil
```
[the remark this answers is not in the document]
Comment (David-T-White) [reply pkg/controller/publisher.go]: yes, pushed related changes

File: pkg/node/node.go (hunk 2 above)
Comment (David-T-White) [reply pkg/node/node.go]: I changed the way this string is built, but did not add a length check at this time. My thinking is we can skip such a check - given the current prefix/name values the only way to get beyond 63 character length would be for a system to have a 30 digit number worth of elements in sasAddresses.

Comment (jskazinski) [reply pkg/node/node.go]: Okay

**Review by jskazinski — APPROVED** [review]

**Pull request merged**